index.d.ts

```diff
@@ -0,0 +1,170 @@
+declare module 'fluture' {
+  interface ResolveFunction<R> {
+    (value: R): void
+  }
+  interface RejectFunction<L> {
+    (error: L): void
+  }
+
+  interface Cancel {
+    (): void
+  }
+
+  export abstract class IFuture<L, R> {
+    fork(reject: RejectFunction<L>, resolve: ResolveFunction<R>): Cancel
+  }
+
+  export function after<L, R>(duration: number, value: R): IFuture<L, R>
+
+  export function and<L, R>(
+    left: IFuture<L, any>,
+    right: IFuture<L, R>
+  ): IFuture<L, R>
+
+  export function ap<L, R, VR>(
+    apply: IFuture<L, (value: VR) => R>,
```

Comment:
Let's have a look at the [documentation for `ap`](https://github.com/fluture-js/Fluture#ap). When you expand the function signature, we see this:

```hs
ap                  :: Apply m => m        (a -> b) -> m        a -> m        b
Future.ap           :: Apply m => m        (a -> b) -> m        a -> m        b
Future.prototype.ap ::            Future e (a -> b) ~> Future e a -> Future e b
```

This shows us that there are *three* distinct versions of the `ap` function. The first and the second are the same, so we can ignore that. It just serves to show that the function lives on the `Future` class *as well as* as a named export (for Static Land compliance).

Let's compare the first and the third.

**The first** is the named export. It's a function with signature:

```hs
ap :: m (a -> b) -> m a -> m b
```

Or in TypeScript:

```ts
function ap<T, A, B>((A) => B, T<A>): T<B>
```

Except that the Hindley-Milner signature has one more part, the leading `Apply m => `. This indicates that, though `m` is a generic type, it must conform to the [`Apply` interface (as defined by Fantasy Land)](https://github.com/fantasyland/fantasy-land#apply).

**The third** is the `ap` *method*. It lives on the prototype of the Future class and is essentially `ap`, but specialized for the Future type:

```hs
ap :: Future e (a -> b) ~> Future e a -> Future e b
```

The `m` has been replaced with `Future e`, and the `Apply m =>` restriction can therefore be omitted. I find it difficult to translate this to TypeScript, let me illustrate:

```ts
class Future<L, R> {
  ap<A, B>(value: Future<L, A>): Future<L, B>
}
```

This TypeScript definition is pretty crummy, as it fails to show where `A` goes, and where `B` comes from. The fact being that `R` in this case must be `(A) => B`. If we would want to improve it, we would have to make use of interfaces such as is done [here](https://github.com/gcanti/fp-ts/blob/master/src/Apply.ts#L17-L19), I think.